include/runtime/collections/rangemap.h

```diff
@@ -0,0 +1,426 @@
+#ifndef RANGEMAP_HEADER_H
+#define RANGEMAP_HEADER_H
+
+#include "RBTree.h"
+
+#include <cassert>
+#include <iostream>
+#include <optional>
+#include <stdexcept>
+
+// Range bounded inclusively below and exlusively above: [start, end).
+// - T : class of values within the range
+template <class T>
+class Range {
+private:
+  T start_;
+  T end_;
+
+public:
+  // Create a new range [start, end).
+  Range(T start, T end)
+      : start_(start)
+      , end_(end) { }
+
+  // Getter for the start of this range.
+  T get_start() const { return start_; }
+
+  // Getter for the end of this range.
+  T get_end() const { return end_; }
+
+  // The following methods define the ordering for objects of class Range.
+  // Operator < is used to subsequently define >, ==, <=, >=, and !=.
+  bool operator<(Range const &other) const {
+    return this->start_ == other.start_ ? this->end_ < other.end_
+                                        : this->start_ < other.start_;
+  }
+
+  bool operator>(Range const &other) const { return other < *this; }
+
+  bool operator==(Range const &other) const {
+    return !(*this < other || other < *this);
+  }
+
+  bool operator<=(Range const &other) const { return !(other < *this); }
+
+  bool operator>=(Range const &other) const { return !(*this < other); }
+
+  bool operator!=(Range const &other) const {
+    return *this < other || other < *this;
+  }
+
+  // Returns true if this range contains k.
+  bool contains(T const &k) const { return k >= start_ && k < end_; }
+
+  // Returns true if this range is empty.
+  bool is_empty() const { return start_ >= end_; }
```

Comment:
`empty` is preferable as previous comments

Reply:
Renamed.